src/main/java/io/reactivex/internal/operators/flowable/FlowableLimit.java

```diff
@@ -0,0 +1,137 @@
+/**
+ * Copyright (c) 2016-present, RxJava Contributors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in
+ * compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is
+ * distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See
+ * the License for the specific language governing permissions and limitations under the License.
+ */
+
+package io.reactivex.internal.operators.flowable;
+
+import java.util.concurrent.atomic.AtomicLong;
+
+import org.reactivestreams.*;
+
+import io.reactivex.*;
+import io.reactivex.annotations.Experimental;
+import io.reactivex.internal.subscriptions.*;
+import io.reactivex.plugins.RxJavaPlugins;
+
+/**
+ * Limits both the total request amount and items received from the upstream.
+ *
+ * @param <T> the source and output value type
+ * @since 2.1.6 - experimental
+ */
+@Experimental
+public final class FlowableLimit<T> extends AbstractFlowableWithUpstream<T, T> {
+
+    final long n;
+
+    public FlowableLimit(Flowable<T> source, long n) {
+        super(source);
+        this.n = n;
+    }
+
+    @Override
+    protected void subscribeActual(Subscriber<? super T> s) {
+        source.subscribe(new LimitSubscriber<T>(s, n));
+    }
+
+    static final class LimitSubscriber<T>
+    extends AtomicLong
+    implements FlowableSubscriber<T>, Subscription {
+
+        private static final long serialVersionUID = 2288246011222124525L;
+
+        final Subscriber<? super T> actual;
+
+        long remaining;
+
+        Subscription upstream;
+
+        LimitSubscriber(Subscriber<? super T> actual, long remaining) {
+            this.actual = actual;
+            this.remaining = remaining;
+            lazySet(remaining);
+        }
+
+        @Override
+        public void onSubscribe(Subscription s) {
+            if (SubscriptionHelper.validate(this.upstream, s)) {
+                if (remaining == 0L) {
+                    s.cancel();
+                    EmptySubscription.complete(actual);
+                } else {
+                    this.upstream = s;
+                    actual.onSubscribe(this);
+                }
+            }
+        }
+
+        @Override
+        public void onNext(T t) {
+            long r = remaining;
```

Comment:
what's the point of local `r`? It looks like you can just do everything directly on `remaining` here

Reply:
I avoid re-reading fields.

Reply:
hm, it's not volatile, so it should be normally in CPU register/cache, but yeah local var has slightly more chances to be there

Although local var creates indirection between read/writes, so overall I think it's neither win nor loss in performance :)

There is interesting question on SO, but answers seem to only compare speed of access to the field when it's already initialized, not taking overall overhead of additional variable https://stackoverflow.com/questions/21613098/java-local-vs-instance-variable-access-speed

Reply:
The JIT may load it into a register but not everything runs on HotSpot.